src/ucsi/lpm/get_connector_status.rs

```diff
@@ -0,0 +1,718 @@
+        };
+
+        Ok(ResponseData {
+            status_change,
+            connect_status,
+            status,
+            reverse_current_protection_status,
+            power_reading,
+        })
+    }
+}
+
+impl From<ResponseData> for [u8; RESPONSE_DATA_LEN] {
+    fn from(data: ResponseData) -> Self {
+        let mut raw = ResponseDataRaw([0; RESPONSE_DATA_LEN as usize]);
+
+        raw.set_status_change(data.status_change.0 .0);
+        raw.set_connect_status(data.connect_status);
+
+        if let Some(status) = data.status {
+            raw.set_power_op_mode(status.power_op_mode as u8);
+            raw.set_power_direction(status.power_direction == PowerRole::Source);
+            raw.set_partner_flags(status.partner_flags as u8);
+            raw.set_partner_type(status.partner_type as u8);
+            if let Some(rdo) = status.rdo {
+                raw.set_rdo(rdo);
+            }
+            if let Some(battery_charging_status) = status.battery_charging_status {
+                raw.set_battery_charging_status(battery_charging_status as u8);
+            }
+            if let Some(bcd_pd_version) = status.bcd_pd_version {
+                raw.set_bcd_pd_version(bcd_pd_version);
+            }
+            raw.set_orientation(status.orientation == PlugOrientation::CC2);
+            raw.set_sink_path_status(status.sink_path_status);
+        }
+
+        raw.set_reverse_current_protection_status(data.reverse_current_protection_status);
+
+        if let Some(power_reading) = data.power_reading {
+            raw.set_power_reading_ready(true);
+            raw.set_current_scale((power_reading.scale_ma / MA5_UNIT) as u8);
+            raw.set_peak_current(power_reading.peak_current_ma / power_reading.scale_ma);
+            raw.set_avg_current(power_reading.avg_current_ma / power_reading.scale_ma);
+            raw.set_voltage_scale((power_reading.scale_mv / MV5_UNIT) as u8);
+            raw.set_voltage_reading(power_reading.voltage_reading_mv / power_reading.scale_mv);
+        } else {
+            raw.set_power_reading_ready(false);
+        }
+
+        raw.0
+    }
+}
+
+impl Encode for ResponseData {
+    fn encode<E: Encoder>(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        <[u8; RESPONSE_DATA_LEN]>::from(*self).encode(encoder)
+    }
+}
+
+impl<Context> Decode<Context> for ResponseData {
+    fn decode<D: Decoder<Context = Context>>(decoder: &mut D) -> Result<Self, DecodeError> {
+        let raw = <[u8; RESPONSE_DATA_LEN]>::decode(decoder)?;
+        ResponseData::try_from(raw).map_err(|e| match e {
+            InvalidResponseData::InvalidPowerOperationMode(err) => DecodeError::from(err),
+            InvalidResponseData::InvalidConnectorPartnerFlags(err) => DecodeError::from(err),
+            InvalidResponseData::InvalidConnectorPartnerType(err) => DecodeError::from(err),
+            InvalidResponseData::InvalidBatteryChargingCapabilityStatus(err) => DecodeError::from(err),
+        })
```

Comment:
You could `impl From<InvalidResponseData> for DecodeError` with this match and shorten this block to
```suggestion
        let data = ResponseData::try_from(raw)?;
        Ok(data)
```